AirCasting/MicrophoneSession/MicrophoneManager.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  microphoneManager.swift
+//  AirCasting
+//
+//  Created by Anna Olak on 11/03/2021.
+//
+
+import Foundation
+import AVFoundation
+import CoreAudio
+import CoreData
+
+class MicrophoneManager: ObservableObject {
+    var context: NSManagedObjectContext {
+        PersistenceController.shared.container.viewContext
+    }
+    var session: Session?
+    var measurementStream: MeasurementStream?
+    
+    private var recorder: AVAudioRecorder!
+    private var levelTimer = Timer()
+    
+    private var locationProvider = LocationProvider()
+    
+    func startRecording(session: Session) {
+        let audioSession = AVAudioSession.sharedInstance()
+        
+        //Check microphone permisson
+        if audioSession.recordPermission != .granted {
+            audioSession.requestRecordPermission { (isGranted) in
+                if !isGranted {
+                    // TODO: pop-up that informs user that we need access to mic with ability do go back to homescreen
+                    fatalError("You must allow audio recording for this demo to work")
+                }
+            }
+        }
+        
+        self.session = session
+        createDBStream()
+        locationProvider.requestLocation()
+        
+        let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
+        
+        let recordSettings: [String: Any] = [
+            AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
+            AVSampleRateKey: 44100.0,
+            AVNumberOfChannelsKey: 1,
+            AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
+        ]
+        
+        do {
+            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
+            try audioSession.setActive(true)
+            try recorder = AVAudioRecorder(url:url, settings: recordSettings)
+            
+        } catch {
+            return
+        }
+        
+        recorder.isMeteringEnabled = true
+        recorder.record()
```

Comment:
I think it might not work properly when it's the first time creating a mic session. 
This code will be executed even if the user is on permission popup, and didn't give us permission yet. Then, the session will be created, `recorder.record()` will be called, etc - all of this before knowing if we have access to the mic. 
I think it would be safer to move this code to the if, like this :
```
        if audioSession.recordPermission != .granted {
 ....... } else {
 self.session = session
        createDBStream()
        locationProvider.requestLocation()
        
        let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
        
        let recordSettings: [String: Any] = [
            AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
            AVSampleRateKey: 44100.0,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
        ]
        
        do {
            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
            try audioSession.setActive(true)
            try recorder = AVAudioRecorder(url:url, settings: recordSettings)
            
        } catch {
            return
        }
        
        recorder.isMeteringEnabled = true
        recorder.record()
        
        levelTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(getMeasurement), userInfo: nil, repeats: true)
}
```

Maybe, you could extra this part of code to a different function, too?

Reply:
This doesn't work because code in the else statement doesn't execute at all if the permission wasn't granted prior to entering this view.